File: Modules/HumanResource/CompanySettings/CompanySettings/CompanySettingsForm.cs
using Serenity.ComponentModel;
using System;
using System.ComponentModel;
using System.Collections.Generic;
using HRMSoftware.AnnualLeavePolicy;
using HRMSoftware.AnnualLeaveJobGradePolicy;
using HRMSoftware.OTJobGradeTime;

namespace HRMSoftware.CompanySettings.Forms;

[FormScript("HumanResource.CompanySettings.CompanySettings")]
[BasedOnRow(typeof(CompanySettingsRow), CheckNames = true)]
public class CompanySettingsForm
{
    [Tab("Company Information")]
    public string CompanyLogo { get; set; }
    public string CompanyRegistrationNumber { get; set; }
    [OneThirdWidth]
    public string CompanyPhone { get; set; }

    [OneThirdWidth]
    public string CompanyName { get; set; }
    [OneThirdWidth]
    public string BasedCountry { get; set; }
    public string CompanyAddress { get; set; }


    public string EPFAccountNumber { get; set; }
    public string BankAccountNumber { get; set; }
    public string SocsoAccountNumber { get; set; }
    public string IncomeTaxAccountNumber { get; set; }
    public string ZakatAccountNumber { get; set; }
    public int BankId { get; set; }


    [Tab("Shift Pattern Information")]

    [Category("You can choose more than one Shift Pattern Method")]
    [OneThirdWidth]
    public bool FixedTime { get; set; }
    [OneThirdWidth]
    public bool FixedHour { get; set; }
    [OneThirdWidth]
    public bool FixedHourFlexiTime { get; set; }
    [Tab("Leave Refresh Method")]

    [Category("Please choose only one Leave Refresh Method")]
    
    [HalfWidth]
    public int RefreshLeaveOnYearOfService { get; set; }
    [HalfWidth]
    public int RefreshLeaveOnSpecificDate { get; set; }
    [HalfWidth]
    public int LeaveRefreshMonth { get; set; }
    [HalfWidth]
    public int LeaveRefreshDay { get; set; }
    
    [Category("Please choose only one Leave Entitlement Method")]
    [QuarterWidth]
    public int EntitleAnnualBasedOnJobGrade { get; set; }
    [QuarterWidth]
    public int EntitleAnnualBasedOnYearOfService { get; set; }


    [Category("Please choose only one Leave Giving Method")]
    [QuarterWidth]
    public int OneOffEntitlementAnnualLeave { get; set; }
    [QuarterWidth]
    public int MonthlyEntitlementAnnualLeave { get; set; }
    [QuarterWidth]
    public int EntitleAnualLeaveBeforeProbationPeriodEnd { get; set; }


    [Tab("Salary Information")]
    [QuarterWidth]
    public int DeductSalaryIfLate { get; set; }
    [QuarterWidth]
    public int DeductSalaryIfEarlyLeaving { get; set; }
    //public DateTime LeaveRefreshDate { get; set; }

    [HalfWidth]
    public int RetireAge { get; set; }

    
    //[Category("Pay Date")]
    public int PayDay { get; set; }
    //public DateTime Paydate { get; set; }
    [Category("Please fill in minimum time in minutes to be considered as Half Day Leave or Absent")]
    [QuarterWidth]
    public int LateArrivalEqualHalfDayLeave { get; set; }
    [QuarterWidth]
    public int LateArrivalEqualFullDayLeave { get; set; }

    [QuarterWidth]
    public int ClockInGracePeriod { get; set; }

    [QuarterWidth]
    public int ClockOutGracePeriod { get; set;}

    [Tab("OT Settings")]
    [Category("Please select Ot Rate Calculation Method")]
    [OneThirdWidth]
    public bool FixedOtRateCalculation { get; set; }
    [OneThirdWidth]
    public bool VariableOtRateCalculation { get; set; }
    [OneThirdWidth]
    public bool OTMonthDays { get; set; }


    
    public float FixedOtRateDenominator { get; set; }
    public string OtCalculationDescription { get; set; }


    [Category("Weekday Rate")]
    public bool WeekdayTwo { get; set; }
    public bool WeekdayOnePointFive { get; set; }
    [Category("Weekend Rate")]
    public bool WeekendTwo { get; set; }
    public bool WeekendOnePointFive { get; set; }
    [Category("Public Holiday Rate")]
    public bool PublicHolidayTwo { get; set; }
    public bool PublicHolidayOnePointFive { get; set; }



    [Category("Please fill in minimum time in minutes to be considered as OT(in minutes)")]
    [OneThirdWidth]
    public int OTMinimumMinute { get; set; }
    [HalfWidth]
    public int MaximumOtMinute { get; set; }

    [Category("Maximum Basic to Entitle for OT Salary, can be left empty")]
    [QuarterWidth]
    public int MaximumBasicSalaryToEntitleForOTPay { get; set; }

    [HalfWidth]
    public int MaximumBasicSalary { get; set; }

    


    [Category("Job Grade OT Entitlement Condition, can be left empty")]
    [QuarterWidth]
    public int MaximumJobGradeToEntitleForOTPay { get; set; }

    [IgnoreName, LabelWidth("OT Maximum Time(hours)"), OTJobGradeTimeEditor]
    public List<OTJobGradeTimeRow> OTJobGradeTime { get; set; }

    [Tab("Penalty calculation settings")]
    [Category("Daily No Paid Leave Rate")]
    [QuarterWidth]
    public bool DailyNplBasedOnWorkingHour { get; set; }
    [QuarterWidth]
    public bool DailyNplBasedOnFixedDenominator { get; set; }
    [QuarterWidth]
    public float FixedNPLDenominator { get; set; }
    [QuarterWidth]
    public bool DailyNplMonthDays { get; set; }
    public string NoPaidLeaveDes { get; set; }
    [Category("Hourly No Paid Leave Rate")]
    [HalfWidth]
    public bool HourlyNoPaidLeaveNPL { get; set; }
    [HalfWidth]
    public bool HourlyNoPaidLeaveAbsent { get; set; }
    public string HourlyNoPaidLeaveDes { get; set; }
    [Category("Daily Absent Rate")]
    [QuarterWidth]
    public bool AbsentBasedOnWorkingHour { get; set; }
    [QuarterWidth]
    public bool AbsentBasedOnFixedDenominator { get; set; }
    [QuarterWidth]
    public float FixedAbsentDenominator { get; set; }
    [QuarterWidth]
    public bool AbsentMonthDays { get; set; }
    public string AbsentDes { get; set; }

    [Category("Early Leaving Rate")]
    [QuarterWidth]
    public bool EarlyLeavingBasedOnWorkingHour { get; set; }
    [QuarterWidth]
    public bool EarlyLeavingBasedOnFixedDenominator { get; set; }
    [QuarterWidth]
    public float FixedEarlyLeavingDenominator { get; set; }
    [QuarterWidth]
    public bool EarlyLeavingMonthDays { get; set; }

    public string EarlyLeavingDes { get; set; }


    [Category("Late Arrival Rate")]
    [QuarterWidth]
    public bool LateArrivalBasedOnWorkingHour { get; set; }
    [QuarterWidth]
    public bool LateArrivalBasedOnFixedDenominator { get; set; }
    [QuarterWidth]
    public float FixedLateArrivalDenominator { get; set; }
    [QuarterWidth]
    public bool LateArrivalMonthDays { get; set; }
    public string LateArrivalDes { get; set; }

    


    [Tab("Working Days Information")]
    public double ProbationPeriod { get; set; }
    public double MonthlyDays { get; set; }

    public bool SundayWeekday { get; set; }
    public bool MondayWeekday { get; set; }
    public bool TuesdayWeekday { get; set; }
    public bool WednesdayWeekday { get; set; }
    public bool ThursdayWeekday { get; set; }
    public bool FridayWeekday { get; set; }
    public bool SaturdayWeekday { get; set; }



    [Tab("Application Approvals")]
    [Category("OT Application")]
    [OneThirdWidth]
    public bool OTApplicationEmployeeApproval { get; set; }
    [OneThirdWidth]
    public bool OTApplicationHrApproval { get; set; }
    [OneThirdWidth]
    public bool OTEitherOne { get; set; }


    
    [Category("Leave Application")]
    [OneThirdWidth]
    public bool LeaveApplicationEmployeeApproval { get; set; }
    [OneThirdWidth]
    public bool LeaveApplicationHrApproval { get; set; }
    [OneThirdWidth]
    public bool LeaveApplicationEitherOne { get; set; }
    [Category("Money Claiming Application")]
    [OneThirdWidth]
    public bool MoneyClaimingEmployeeApproval { get; set; }
    [OneThirdWidth]
    public bool MoneyClaimingHrApproval { get; set; }
    [OneThirdWidth]
    public bool MoneyClaimingEitherOne { get; set; }
    
}

File: Modules/ServerTypes/CompanySettings/CompanySettingsForm.ts
import { ImageUploadEditor, StringEditor, LookupEditor, BooleanEditor, IntegerEditor, DecimalEditor, PrefixedContext } from "@serenity-is/corelib";
import { OTJobGradeTimeEditor } from "@/HumanResource/OTJobGradeTime/OTJobGradeTime/OTJobGradeTimeEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface CompanySettingsForm {
    CompanyLogo: ImageUploadEditor;
    CompanyRegistrationNumber: StringEditor;
    CompanyPhone: StringEditor;
    CompanyName: StringEditor;
    BasedCountry: StringEditor;
    CompanyAddress: StringEditor;
    EPFAccountNumber: StringEditor;
    BankAccountNumber: StringEditor;
    SocsoAccountNumber: StringEditor;
    IncomeTaxAccountNumber: StringEditor;
    ZakatAccountNumber: StringEditor;
    BankId: LookupEditor;
    FixedTime: BooleanEditor;
    FixedHour: BooleanEditor;
    FixedHourFlexiTime: BooleanEditor;
    RefreshLeaveOnYearOfService: BooleanEditor;
    RefreshLeaveOnSpecificDate: BooleanEditor;
    LeaveRefreshMonth: IntegerEditor;
    LeaveRefreshDay: IntegerEditor;
    EntitleAnnualBasedOnJobGrade: BooleanEditor;
    EntitleAnnualBasedOnYearOfService: BooleanEditor;
    OneOffEntitlementAnnualLeave: BooleanEditor;
    MonthlyEntitlementAnnualLeave: BooleanEditor;
    EntitleAnualLeaveBeforeProbationPeriodEnd: BooleanEditor;
    DeductSalaryIfLate: BooleanEditor;
    DeductSalaryIfEarlyLeaving: BooleanEditor;
    RetireAge: IntegerEditor;
    PayDay: IntegerEditor;
    LateArrivalEqualHalfDayLeave: IntegerEditor;
    LateArrivalEqualFullDayLeave: IntegerEditor;
    ClockInGracePeriod: IntegerEditor;
    ClockOutGracePeriod: IntegerEditor;
    FixedOtRateCalculation: BooleanEditor;
    VariableOtRateCalculation: BooleanEditor;
    OTMonthDays: BooleanEditor;
    FixedOtRateDenominator: DecimalEditor;
    OtCalculationDescription: StringEditor;
    WeekdayTwo: BooleanEditor;
    WeekdayOnePointFive: BooleanEditor;
    WeekendTwo: BooleanEditor;
    WeekendOnePointFive: BooleanEditor;
    PublicHolidayTwo: BooleanEditor;
    PublicHolidayOnePointFive: BooleanEditor;
    OTMinimumMinute: IntegerEditor;
    MaximumOtMinute: IntegerEditor;
    MaximumBasicSalaryToEntitleForOTPay: BooleanEditor;
    MaximumBasicSalary: IntegerEditor;
    MaximumJobGradeToEntitleForOTPay: BooleanEditor;
    OTJobGradeTime: OTJobGradeTimeEditor;
    DailyNplBasedOnWorkingHour: BooleanEditor;
    DailyNplBasedOnFixedDenominator: BooleanEditor;
    FixedNPLDenominator: DecimalEditor;
    DailyNplMonthDays: BooleanEditor;
    NoPaidLeaveDes: StringEditor;
    HourlyNoPaidLeaveNPL: BooleanEditor;
    HourlyNoPaidLeaveAbsent: BooleanEditor;
    HourlyNoPaidLeaveDes: StringEditor;
    AbsentBasedOnWorkingHour: BooleanEditor;
    AbsentBasedOnFixedDenominator: BooleanEditor;
    FixedAbsentDenominator: DecimalEditor;
    AbsentMonthDays: BooleanEditor;
    AbsentDes: StringEditor;
    EarlyLeavingBasedOnWorkingHour: BooleanEditor;
    EarlyLeavingBasedOnFixedDenominator: BooleanEditor;
    FixedEarlyLeavingDenominator: DecimalEditor;
    EarlyLeavingMonthDays: BooleanEditor;
    EarlyLeavingDes: StringEditor;
    LateArrivalBasedOnWorkingHour: BooleanEditor;
    LateArrivalBasedOnFixedDenominator: BooleanEditor;
    FixedLateArrivalDenominator: DecimalEditor;
    LateArrivalMonthDays: BooleanEditor;
    LateArrivalDes: StringEditor;
    ProbationPeriod: DecimalEditor;
    MonthlyDays: DecimalEditor;
    SundayWeekday: BooleanEditor;
    MondayWeekday: BooleanEditor;
    TuesdayWeekday: BooleanEditor;
    WednesdayWeekday: BooleanEditor;
    ThursdayWeekday: BooleanEditor;
    FridayWeekday: BooleanEditor;
    SaturdayWeekday: BooleanEditor;
    OTApplicationEmployeeApproval: BooleanEditor;
    OTApplicationHrApproval: BooleanEditor;
    OTEitherOne: BooleanEditor;
    LeaveApplicationEmployeeApproval: BooleanEditor;
    LeaveApplicationHrApproval: BooleanEditor;
    LeaveApplicationEitherOne: BooleanEditor;
    MoneyClaimingEmployeeApproval: BooleanEditor;
    MoneyClaimingHrApproval: BooleanEditor;
    MoneyClaimingEitherOne: BooleanEditor;
}

export class CompanySettingsForm extends PrefixedContext {
    static formKey = 'HumanResource.CompanySettings.CompanySettings';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!CompanySettingsForm.init)  {
            CompanySettingsForm.init = true;

            var w0 = ImageUploadEditor;
            var w1 = StringEditor;
            var w2 = LookupEditor;
            var w3 = BooleanEditor;
            var w4 = IntegerEditor;
            var w5 = DecimalEditor;
            var w6 = OTJobGradeTimeEditor;

            initFormType(CompanySettingsForm, [
                'CompanyLogo', w0,
                'CompanyRegistrationNumber', w1,
                'CompanyPhone', w1,
                'CompanyName', w1,
                'BasedCountry', w1,
                'CompanyAddress', w1,
                'EPFAccountNumber', w1,
                'BankAccountNumber', w1,
                'SocsoAccountNumber', w1,
                'IncomeTaxAccountNumber', w1,
                'ZakatAccountNumber', w1,
                'BankId', w2,
                'FixedTime', w3,
                'FixedHour', w3,
                'FixedHourFlexiTime', w3,
                'RefreshLeaveOnYearOfService', w3,
                'RefreshLeaveOnSpecificDate', w3,
                'LeaveRefreshMonth', w4,
                'LeaveRefreshDay', w4,
                'EntitleAnnualBasedOnJobGrade', w3,
                'EntitleAnnualBasedOnYearOfService', w3,
                'OneOffEntitlementAnnualLeave', w3,
                'MonthlyEntitlementAnnualLeave', w3,
                'EntitleAnualLeaveBeforeProbationPeriodEnd', w3,
                'DeductSalaryIfLate', w3,
                'DeductSalaryIfEarlyLeaving', w3,
                'RetireAge', w4,
                'PayDay', w4,
                'LateArrivalEqualHalfDayLeave', w4,
                'LateArrivalEqualFullDayLeave', w4,
                'ClockInGracePeriod', w4,
                'ClockOutGracePeriod', w4,
                'FixedOtRateCalculation', w3,
                'VariableOtRateCalculation', w3,
                'OTMonthDays', w3,
                'FixedOtRateDenominator', w5,
                'OtCalculationDescription', w1,
                'WeekdayTwo', w3,
                'WeekdayOnePointFive', w3,
                'WeekendTwo', w3,
                'WeekendOnePointFive', w3,
                'PublicHolidayTwo', w3,
                'PublicHolidayOnePointFive', w3,
                'OTMinimumMinute', w4,
                'MaximumOtMinute', w4,
                'MaximumBasicSalaryToEntitleForOTPay', w3,
                'MaximumBasicSalary', w4,
                'MaximumJobGradeToEntitleForOTPay', w3,
                'OTJobGradeTime', w6,
                'DailyNplBasedOnWorkingHour', w3,
                'DailyNplBasedOnFixedDenominator', w3,
                'FixedNPLDenominator', w5,
                'DailyNplMonthDays', w3,
                'NoPaidLeaveDes', w1,
                'HourlyNoPaidLeaveNPL', w3,
                'HourlyNoPaidLeaveAbsent', w3,
                'HourlyNoPaidLeaveDes', w1,
                'AbsentBasedOnWorkingHour', w3,
                'AbsentBasedOnFixedDenominator', w3,
                'FixedAbsentDenominator', w5,
                'AbsentMonthDays', w3,
                'AbsentDes', w1,
                'EarlyLeavingBasedOnWorkingHour', w3,
                'EarlyLeavingBasedOnFixedDenominator', w3,
                'FixedEarlyLeavingDenominator', w5,
                'EarlyLeavingMonthDays', w3,
                'EarlyLeavingDes', w1,
                'LateArrivalBasedOnWorkingHour', w3,
                'LateArrivalBasedOnFixedDenominator', w3,
                'FixedLateArrivalDenominator', w5,
                'LateArrivalMonthDays', w3,
                'LateArrivalDes', w1,
                'ProbationPeriod', w5,
                'MonthlyDays', w5,
                'SundayWeekday', w3,
                'MondayWeekday', w3,
                'TuesdayWeekday', w3,
                'WednesdayWeekday', w3,
                'ThursdayWeekday', w3,
                'FridayWeekday', w3,
                'SaturdayWeekday', w3,
                'OTApplicationEmployeeApproval', w3,
                'OTApplicationHrApproval', w3,
                'OTEitherOne', w3,
                'LeaveApplicationEmployeeApproval', w3,
                'LeaveApplicationHrApproval', w3,
                'LeaveApplicationEitherOne', w3,
                'MoneyClaimingEmployeeApproval', w3,
                'MoneyClaimingHrApproval', w3,
                'MoneyClaimingEitherOne', w3
            ]);
        }
    }
}